### PR TITLE
Make memory hooks work with static libcudart, by using bistro hooks for cuda driver API

### LINKED ISSUE #3210: UCX memory hooks with static linking
I recently ran into an issue with an application, compiled against OpenMPI with UCX support, that manifested itself as a segfault during GPU aware MPI function calls(`MPI_Isend` in this particular case). The issue was caused by the application statically linking `libcudart`, which caused the UCX `cudaMalloc` memory hook to not be registered correctly, leading UCX to treat the GPU memory as CPU memory. Is it possible for UCX to detect such situations and handle them more gracefully? If that's not so easy perhaps this requirement could be better documented?

### PR DESCRIPTION
**Pull request by yosefe**
# Why
Support Cuda memory allocate/release hooks also for applications which are linked to libcudart statically, or use directly the driver API.
Fixed #3210

# How
Use binary instrumentation (bistro) hooks for Cuda Driver APIs, which work regardless of how a function is called (from .plt or by function pointer). 
- Before this PR, if an application was statically linked to libcudart, we could not install hooks on Cuda memory events: cuda runtime calls (e.g cudaFree) are not going through .plt, and libcudart is calling driver API (e.g cuMemFree) by function pointer.
- After this PR, the calls from cudart to cuda driver API can be intercepter.
 
If any other library is intercepting the same function by .plt hooks, it can work alongside with bistro hooks. However, if the other library is also using binary patching, in will probably be incompatible with the UCX hooks.

Binary instrumentation hook method:
1. Copy minimal number of bytes of target function prologue to a newly allocated buffer with execute permissions.
2. After these code bytes, place an indirect jump instruction to the rest of the target function code.
     > If there are not enough position independent instructions, such as `push <reg>` or `sub <imm>, <reg>`, to accommodate 
     > for the indirect jump (3), this method will fail.
2. Place an indirect jump in the target function start to the UCM replacement
3. UCM replacement will treat the executable buffer (1) as the "original" function which should be used from the replacement (2)
```
memory allocate/release
in the application
      +
      |
      |       target function
      |       (e.g cuMemAlloc)               UCM replacement
      +----->-----------------+              (e.g ucm_cuMemAlloc)
             |jmp to UCM hook +-------------->+-----------+
             |                | (3)           | dispatch  +-------------> [ rcache ]
             |                |               | UCM event |
             +----------------+               |           |
             | unmodified     +<---+          +-----------+
             | code of target |    |          | call orig | (4)
             | function       |    |      +---+ function  |
             |                |    |      |   +-----------+
             |                |    |      |
             |                |    |      |   executable buffer
             |                |    |      |   (e.g ucm_orig_cuMemAlloc)
             |                |    |      +-->+-----------+
             |                |    |          | copy of   | (1)
             |                |    |          | function  |
             |                |    |          | prologue  |
             |                |    |          +-----------+
             |                |    +----------+ jump back | (2)
             +----------------+               +-----------+

```

### CODE REVIEW

File: src/ucm/bistro/bistro_x86_64.c [hunk 1]
```diff
@@ -14,19 +14,153 @@
...
+        /* check for REX prefix */
+        if ((opcode & 0xF0) == 0x40) {
+            rex    = opcode;
+            opcode = *(uint8_t*)UCS_PTR_BYTE_OFFSET(func, offset++);
+        } else {
+            rex = 0;
+        }
+
+        /* check the opcode */
+        if (((rex == 0) || rex == 0x41) && ((opcode & 0xF0) == 0x50)) {
```
Comment (hoopoepg) [inline src/ucm/bistro/bistro_x86_64.c:64]: ```((rex == 0) || rex == 0x41)``` is this correct?

File: src/ucm/bistro/bistro_x86_64.c [hunk 2]
```diff
@@ -14,19 +14,153 @@
...
+            }
+        } else if ((rex == 0x48) && (opcode == 0x89)) {
+            /* mov %rsp, %rbp */
+            modrm = *(uint8_t*)UCS_PTR_BYTE_OFFSET(func, offset++);
+            if (modrm == 0xE5) {
+                continue;
+            }
+        }
+
+        /* unsupported instruction - bail */
```
Comment (hoopoepg) [inline src/ucm/bistro/bistro_x86_64.c:83]: bail? not fail?

File: src/ucm/bistro/bistro_x86_64.c [hunk 3]
```diff
@@ -14,19 +14,153 @@
...
+static void *ucm_bistro_allocate_code(size_t size)
+{
+    static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+    void *page                   = MAP_FAILED;
+    size_t offset                = 0;
+    void *result;
+
+    pthread_mutex_lock(&mutex);
+
+    if (page == MAP_FAILED) {
```
Comment (hoopoepg) [inline src/ucm/bistro/bistro_x86_64.c:99]: this condition is always true

File: src/ucm/bistro/bistro_x86_64.c [hunk 4]
```diff
@@ -14,19 +14,153 @@
...
+        page = mmap(NULL, ucm_get_page_size(),
+                    PROT_READ | PROT_WRITE | PROT_EXEC,
+                    MAP_PRIVATE | MAP_ANONYMOUS, 0, 0);
+        if (page == MAP_FAILED) {
+            result = NULL;
+            goto out;
+        }
+    }
+
+    if (size + offset > ucm_get_page_size()) {
```
Comment (hoopoepg) [inline src/ucm/bistro/bistro_x86_64.c:110]: offset is 0 here... and wrap ```size + offset``` by ()

File: src/ucm/bistro/bistro_x86_64.c [hunk 5]
```diff
@@ -14,19 +14,153 @@
...
+        result = NULL;
+        goto out;
+    }
+
+    /* allocate space in the page */
+    result  = UCS_PTR_BYTE_OFFSET(page, offset);
+    offset += size;
```
Comment (hoopoepg) [inline src/ucm/bistro/bistro_x86_64.c:117]: it seems body of this function was in loop :)

Comment (yosefe) [reply src/ucm/bistro/bistro_x86_64.c:117]: forgot to add 'static'

File: src/ucm/bistro/bistro_x86_64.c (hunk 1 above)
Comment (yosefe) [reply src/ucm/bistro/bistro_x86_64.c:63]: rex == 0 is pushing IA32 register and rex==0x41 is pushing rN register

File: src/ucm/bistro/bistro_x86_64.c (hunk 2 above)
Comment (yosefe) [reply src/ucm/bistro/bistro_x86_64.c:82]: yes, "bail" means exit

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
@Akshay-Venkatesh @bureddy can you pls take a look?

**Comment by yosefe** [comment]
@bureddy fixed

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> @Akshay-Venkatesh @bureddy can you pls take a look?

Some high-level questions/requests:
 - Can you please explain how/why this solves the static linking issue? "How" section in the PR seems to describe the low-level mechanism of binary instrumentation. Specifying current limitations with cudart static linking (such as use of function pointers in cudart), how current hooks in ucm/cuda fails to intercept when cudart static is used, and how this PR addresses issues could be useful for future reference. 
 - Should we document any limitations with current PR (such as memory allocated before ucp_init)?
 - Does this PR address the case when there are other libraries intercept cu* functions?

**Comment by yosefe** [comment]
> * Can you please explain how/why this solves the static linking issue? "How" section in the PR seems to describe the low-level mechanism of binary instrumentation. Specifying current limitations with cudart static linking (such as use of function pointers in cudart), how current hooks in ucm/cuda fails to intercept when cudart static is used, and how this PR addresses issues could be useful for future reference.

Added

> * Should we document any limitations with current PR (such as memory allocated before ucp_init)?

Memory before ucp_init is not a problem: we set suspicious regions to UNKNOWN and force detection during runtime

> * Does this PR address the case when there are other libraries intercept cu* functions?

Added note: if other library is hooking .plt it would work, if using bistro - no

File: src/ucm/cuda/cudamem.c [hunk 6]
```diff
@@ -6,392 +6,263 @@
...
 
-    ucm_event_leave();
-    return ret;
-}
+        status = UCS_ERR_UNSUPPORTED;
 
-cudaError_t ucm_cudaHostGetDevicePointer(void **pDevice, void *pHost, unsigned int flags)
-{
-    cudaError_t ret;
+        if (ucm_mmap_hook_mode() == UCM_MMAP_HOOK_BISTRO) {
```
Comment (bureddy) [inline src/ucm/cuda/cudamem.c:225]: @yosefe  should we detach cuda hooks from ucm_mmap_hook_mode() flag? may be we have a case where we need to to disable malloc bistro hooks but not cuda hooks ( for ex: on summit with spectrum mpi).
how about controlling cuda hooks with new flag "ucm_cuda_hook_mode: and "remove ucm_global_opts.enable_cuda_reloc" ?

File: src/ucm/cuda/cudamem.c [hunk 7]
```diff
@@ -261,7 +266,7 @@ static ucs_status_t ucm_cudamem_install(int events)
         goto out;
     }
 
-    if (!ucm_global_opts.enable_cuda_reloc) {
+    if (ucm_cuda_hook_mode() == UCM_MMAP_HOOK_NONE) {
         ucm_debug("cudamem relocations are disabled by configuration");
```
Comment (bureddy) [inline src/ucm/cuda/cudamem.c:270]: s/relocations/hooks?

File: src/ucs/config/ucm_opts.c [hunk 8]
```diff
@@ -68,10 +69,21 @@ static ucs_config_field_t ucm_global_config_table[] = {
    "which would use the original implementation and not ours.",
    ucs_offsetof(ucm_global_config_t, enable_malloc_reloc), UCS_CONFIG_TYPE_BOOL},
 
+  {"CUDA_HOOK_MODE", UCM_DEFAULT_HOOK_MODE_STR,
+   "Cuda memory hook mode\n"
+   " none   - Don't set Cuda hooks.\n"
+   " reloc  - Use ELF relocation table to set hooks.\n"
+#if UCM_BISTRO_HOOKS
+   " bistro - Use binary instrumentation to set hooks. In this mode, calls to\n"
+   "          statically a statically Cuda runtime can be hooked on the the\n"
```
Comment (bureddy) [inline src/ucs/config/ucm_opts.c:78]: seems wording is not clear

File: src/ucm/cuda/cudamem.c [hunk 9]
```diff
@@ -402,17 +278,23 @@ static ucs_status_t ucm_cudamem_install(int events)
...
+    if (status != UCS_OK) {
+        ucm_warn("failed to install cuda memory hooks on driver API");
+    } else if (!used_reloc) {
+        ucm_debug("installed cuda memory bistro hooks on driver API");
+        ucm_cudamem_installed = 1;
+    } else if (status == UCS_OK) {
+        status = ucm_cuda_install_hooks(ucm_cuda_runtime_funcs, &used_reloc);
+        if (status == UCS_OK) {
+            ucm_debug("installed cuda memory %s hooks on runtime API",
+                      used_reloc ? "reloc" : "bistro");
```
Comment (bureddy) [inline src/ucm/cuda/cudamem.c:291]: seems "ucm_cuda_install_hooks" can do bistro for some funcs if it is possible and reloc for others. is this debug msg accurate?

**Review by bureddy — APPROVED** [review]

**Pull request merged**